src/cls/rbd/cls_rbd_types.cc

```diff
@@ -145,26 +147,34 @@ void MirrorImage::decode(bufferlist::const_iterator &it) {
     decode(int_mode, it);
     mode = static_cast<MirrorImageMode>(int_mode);
   }
+  if (struct_v >= 3) {
+    decode(group_spec, it);
+  }
   DECODE_FINISH(it);
 }
 
 void MirrorImage::dump(Formatter *f) const {
   f->dump_stream("mode") << mode;
   f->dump_string("global_image_id", global_image_id);
+  if (group_spec.is_valid()) {
+    f->open_object_section("group");
+    group_spec.dump(f);
+    f->close_section();
+  }
   f->dump_stream("state") << state;
 }
 
 void MirrorImage::generate_test_instances(std::list<MirrorImage*> &o) {
   o.push_back(new MirrorImage());
-  o.push_back(new MirrorImage(MIRROR_IMAGE_MODE_JOURNAL, "uuid-123",
+  o.push_back(new MirrorImage(MIRROR_IMAGE_MODE_JOURNAL, "uuid-123", {},
                               MIRROR_IMAGE_STATE_ENABLED));
-  o.push_back(new MirrorImage(MIRROR_IMAGE_MODE_SNAPSHOT, "uuid-abc",
+  o.push_back(new MirrorImage(MIRROR_IMAGE_MODE_SNAPSHOT, "uuid-abc", {"g1", 1},
                               MIRROR_IMAGE_STATE_DISABLING));
 }
 
 bool MirrorImage::operator==(const MirrorImage &rhs) const {
   return mode == rhs.mode && global_image_id == rhs.global_image_id &&
-         state == rhs.state;
+         group_spec == rhs.group_spec && state == rhs.state;
```

Comment:
Should this check for a valid group_spec before comparing?

Reply:
Same as for `operator<`, I don't think so.  All group specs, including one that signifies that an image isn't part of any group, are valid.  Perhaps `is_valid()` should be renamed to `is_empty()`?

Reply:
@idryomov , this comment no longer holds as we intend to remove the GroupSpec from the MirrorImage, correct?

Reply:
Correct.